deface/deface.py

```diff
@@ -102,6 +128,106 @@ def cam_read_iter(reader):
         yield reader.get_next_data()
 
 
+def has_overlap(det, other):
+    # In openCV the coordinator start at the top left corner of the image
+    x1, y1, x2, y2, _ = det
+    X1, Y1, X2, Y2, _ = other
+    h_overlaps = (x1 <= X2) and (x2 >= X1)
+    v_overlaps = (y1 <= Y2) and (y2 >= Y1)
+    return h_overlaps and v_overlaps
+
+
+def has_overlap_with_group(det, group):
+    for other in group:
+        if has_overlap(det, other):
+            return True
+    return False
+
+
+def group_det_by_overlap(dets):
+    groups = [[]]
+    ordered_dets = sorted(dets, key=lambda x: (x[0], x[1]))
+    # add to groups the det that have overlap with others
+    for det in ordered_dets:
+        if not groups[-1]:
+            groups[-1].append(det)
+            continue
+        if has_overlap_with_group(det, groups[-1]):
+            groups[-1].append(det)
+        else:
+            groups.append([det])
+    return groups
+
+
+def get_group_centeroid(groups):
+    new_dets = []
+    # The box of a group has centroid of weighted average centroids (by area) and
+    # and the width, height and score are the max of all boxes in the group
+    for group in groups:
+        if not group:
+            continue
+        elif len(group) == 1:
+            new_dets.append(group[0])
+        else:
+            x1s = [x[0] for x in group]
+            y1s = [x[1] for x in group]
+            x2s = [x[2] for x in group]
+            y2s = [x[3] for x in group]
+
+            widths = [x2 - x1 for x2, x1 in zip(x2s, x1s)]
+            heights = [y2 - y1 for y2, y1 in zip(y2s, y1s)]
+
+            areas = [w * h for w, h in zip(widths, heights)]
+            max_w = max(widths)
+            max_h = max(heights)
+
+            x_centroids = [np.rint((x1 + x2) / 2) for x1, x2 in zip(x1s, x2s)]
+            y_centroids = [np.rint((y1 + y2) / 2) for y1, y2 in zip(y1s, y2s)]
+            group_x_centroid = np.average(x_centroids, weights=areas)
+            group_y_centroid = np.average(y_centroids, weights=areas)
+
+            group_x1 = max([min(x1s), np.floor(group_x_centroid - max_w / 2)])
+            group_y1 = max([min(y1s), np.floor(group_y_centroid - max_h / 2)])
+            group_x2 = min([max(x2s), np.floor(group_x_centroid + max_w / 2)])
+            group_y2 = min([max(y2s), np.floor(group_y_centroid + max_h / 2)])
+            group_score = max([x[4] for x in group])
+
+            new_dets.append([group_x1, group_y1, group_x2, group_y2, group_score])
+    return np.asarray(new_dets, dtype=np.float32)
```

Comment:
Same comment as below about `new_dets` being an array of array, not an array of floats.

Reply:
Turn out the better practice with numpy is using the list and convert to numpy array later according to [here](https://stackoverflow.com/a/22394181) so I will keep this conversion.